proc/proc.go

```diff
@@ -388,65 +407,29 @@ func (dbp *Process) pickCurrentThread(trapthread *Thread) error {
 // Step will continue until another source line is reached.
 // Will step into functions.
 func (dbp *Process) Step() (err error) {
-	if dbp.SelectedGoroutine != nil && dbp.SelectedGoroutine.thread == nil {
-		// Step called on parked goroutine
-		return dbp.stepToPC(dbp.SelectedGoroutine.PC)
+	if dbp.exited {
+		return &ProcessExitedError{}
 	}
-	fn := func() error {
-		var nloc *Location
-		th := dbp.CurrentThread
-		loc, err := th.Location()
-		if err != nil {
-			return err
-		}
-		for {
-			pc, err := dbp.CurrentThread.PC()
-			if err != nil {
-				return err
-			}
-			text, err := dbp.CurrentThread.Disassemble(pc, pc+maxInstructionLength, true)
-			if err == nil && len(text) > 0 && text[0].IsCall() && text[0].DestLoc != nil && text[0].DestLoc.Fn != nil {
-				fn := text[0].DestLoc.Fn
-				// Ensure PC and Entry match, otherwise StepInto is likely to set
-				// its breakpoint before DestLoc.PC and hence run too far ahead.
-				// Calls to runtime.duffzero and duffcopy have this problem.
-				if fn.Entry == text[0].DestLoc.PC {
-					return dbp.StepInto(fn)
-				}
-			}
-
-			err = dbp.CurrentThread.StepInstruction()
-			if err != nil {
-				return err
-			}
-			nloc, err = th.Location()
-			if err != nil {
-				return err
-			}
-			if nloc.File != loc.File {
-				return nil
-			}
-			if nloc.File == loc.File && nloc.Line != loc.Line {
-				return nil
-			}
+	for i := range dbp.Breakpoints {
+		if dbp.Breakpoints[i].Kind != NormalBreakpoint {
+			return fmt.Errorf("next while nexting")
 		}
 	}
-	return dbp.run(fn)
-}
 
-// StepInto sets a temp breakpoint after the prologue of fn and calls Continue
-func (dbp *Process) StepInto(fn *gosym.Func) error {
-	for i := range dbp.Breakpoints {
-		if dbp.Breakpoints[i].Temp {
-			return fmt.Errorf("next while nexting")
+	if err = dbp.next(true); err != nil {
+		switch err.(type) {
+		case ThreadBlockedError, NoReturnAddr: // Noop
+		default:
+			dbp.ClearTempBreakpoints()
+			return
 		}
 	}
-	pc, _ := dbp.FirstPCAfterPrologue(fn, false)
-	return dbp.stepToPC(pc)
+
+	return dbp.Continue()
 }
 
 func (dbp *Process) stepToPC(pc uint64) error {
-	if _, err := dbp.SetTempBreakpoint(pc, sameGoroutineCondition(dbp.SelectedGoroutine)); err != nil {
+	if _, err := dbp.SetTempBreakpoint(pc, NextBreakpoint, sameGoroutineCondition(dbp.SelectedGoroutine)); err != nil {
```

Comment:
Agreed. What do you think of this signature:
```
func (dbp *Process) SetBreakpoint(addr uint64, kind BreakpointKind, cond ast.Expr) (*Breakpoint, error)
```

The counters are handled by `(*Process).setBreakpoint` anyway, that's not a problem. I would do this on another commit however. I'd do it in a separate PR, however.